// @abp/ng.core/esm2022/lib/services/internet-connection-service.mjs
import { DOCUMENT } from '@angular/common';
import { Injectable, computed, inject, signal } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class InternetConnectionService {
    constructor() {
        this.document = inject(DOCUMENT);
        this.window = this.document.defaultView;
        this.navigator = this.window.navigator;
        this.status$ = new BehaviorSubject(this.navigator.onLine);
        this.status = signal(this.navigator.onLine);
        this.networkStatus = computed(() => this.status());
        this.window.addEventListener('offline', () => this.setStatus(false));
        this.window.addEventListener('online', () => this.setStatus(true));
    }
    setStatus(val) {
        this.status.set(val);
        this.status$.next(val);
    }
    get networkStatus$() {
        return this.status$.asObservable();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: InternetConnectionService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: InternetConnectionService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: InternetConnectionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [] });
