// @abp/ng.core/esm2022/lib/services/title-strategy.service.mjs
import { Injectable, effect, inject } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { TitleStrategy } from '@angular/router';
import { Title } from '@angular/platform-browser';
import { LocalizationService } from './localization.service';
import { DISABLE_PROJECT_NAME } from '../tokens';
import * as i0 from "@angular/core";
export class AbpTitleStrategy extends TitleStrategy {
    constructor() {
        super();
        this.title = inject(Title);
        this.localizationService = inject(LocalizationService);
        this.disableProjectName = inject(DISABLE_PROJECT_NAME, { optional: true });
        this.langugageChange = toSignal(this.localizationService.languageChange$);
        effect(() => {
            if (this.langugageChange()) {
                this.updateTitle(this.routerState);
            }
        });
    }
    updateTitle(routerState) {
        this.routerState = routerState;
        const title = this.buildTitle(routerState);
        const projectName = this.localizationService.instant({
            key: '::AppName',
            defaultValue: 'MyProjectName',
        });
        if (!title) {
            return this.title.setTitle(projectName);
        }
        let localizedText = this.localizationService.instant({ key: title, defaultValue: title });
        if (!this.disableProjectName) {
            localizedText += ` | ${projectName}`;
        }
        this.title.setTitle(localizedText);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AbpTitleStrategy, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AbpTitleStrategy, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AbpTitleStrategy, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [] });
