// @abp/ng.core/esm2022/lib/directives/autofocus.directive.mjs
import { Directive, ElementRef, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class AutofocusDirective {
    set delay(val) {
        this._delay = Number(val) || 0;
    }
    get delay() {
        return this._delay;
    }
    constructor(elRef) {
        this.elRef = elRef;
        this._delay = 0;
    }
    ngAfterViewInit() {
        setTimeout(() => this.elRef.nativeElement.focus(), this.delay);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AutofocusDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.5", type: AutofocusDirective, isStandalone: true, selector: "[autofocus]", inputs: { delay: ["autofocus", "delay"] }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AutofocusDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[autofocus]',
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }], propDecorators: { delay: [{
                type: Input,
                args: ['autofocus']
            }] } });
