// @abp/ng.core/esm2022/lib/guards/permission.guard.mjs
import { Injectable, inject } from '@angular/core';
import { Router, } from '@angular/router';
import { of } from 'rxjs';
import { tap } from 'rxjs/operators';
import { AuthService } from '../abstracts';
import { findRoute, getRoutePath } from '../utils/route-utils';
import { RoutesService, PermissionService, HttpErrorReporterService } from '../services';
import * as i0 from "@angular/core";
/**
 * @deprecated Use `permissionGuard` *function* instead.
 */
export class PermissionGuard {
    constructor() {
        this.router = inject(Router);
        this.routesService = inject(RoutesService);
        this.authService = inject(AuthService);
        this.permissionService = inject(PermissionService);
        this.httpErrorReporter = inject(HttpErrorReporterService);
    }
    canActivate(route, state) {
        let { requiredPolicy } = route.data || {};
        if (!requiredPolicy) {
            const routeFound = findRoute(this.routesService, getRoutePath(this.router, state.url));
            requiredPolicy = routeFound?.requiredPolicy;
        }
        if (!requiredPolicy)
            return of(true);
        return this.permissionService.getGrantedPolicy$(requiredPolicy).pipe(tap(access => {
            if (!access && this.authService.isAuthenticated) {
                this.httpErrorReporter.reportError({ status: 403 });
            }
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: PermissionGuard, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: PermissionGuard, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: PermissionGuard, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
export const permissionGuard = (route, state) => {
    const router = inject(Router);
    const routesService = inject(RoutesService);
    const authService = inject(AuthService);
    const permissionService = inject(PermissionService);
    const httpErrorReporter = inject(HttpErrorReporterService);
    let { requiredPolicy } = route.data || {};
    if (!requiredPolicy) {
        const routeFound = findRoute(routesService, getRoutePath(router, state.url));
        requiredPolicy = routeFound?.requiredPolicy;
    }
    if (!requiredPolicy)
        return of(true);
    return permissionService.getGrantedPolicy$(requiredPolicy).pipe(tap(access => {
        if (!access && authService.isAuthenticated) {
            httpErrorReporter.reportError({ status: 403 });
        }
    }));
};
