// @abp/ng.core/esm2022/lib/abstracts/auth.guard.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @deprecated Use `authGuard` *function* instead.
 */
export class AuthGuard {
    canActivate() {
        console.error('You should add @abp/ng-oauth packages or create your own auth packages.');
        return false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AuthGuard, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AuthGuard, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.5", ngImport: i0, type: AuthGuard, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
export const authGuard = () => {
    console.error('You should add @abp/ng-oauth packages or create your own auth packages.');
    return false;
};
